src/api/app/jobs/stage_requests_job.rb

```diff
@@ -0,0 +1,27 @@
+class StageRequestsJob < ApplicationJob
+  def perform(staging_project, request_ids, user)
+    unassigned = []
+    User.current = user
+    requests = BsRequest.find(request_ids)
+    requests.each do |request|
+      action = request.bs_request_actions.first # we only care about the first one
+      if action.is_submit?
+        BranchPackage.new(
+          target_project: staging_project.name,
+          target_package: action.target_package,
+          project: action.source_project,
+          package: action.source_package,
+          extend_package_names: false
+        ).branch
+      elsif request.is_delete?
+        # TODO: implement delete requests
+      end
+    rescue APIError => api_error
+      Airbrake.notify(api_error, bs_request: request.number)
+      unassigned << request
+    end
+    # If something goes wrong with branching the package we unassign it again
```

Comment:
Is that something that really can happen? That sounds to me like this would be a bug in the branching command then.

Reply:
> Is that something that really can happen? That sounds to me like this would be a bug in the branching command then.

I don't now exactly about what you are talking about, if you are talking about the ```rescue``` just look how many times we call ```raise``` through the ```branch_package.rb```

Reply:
To give a little more context: Like @vpereira mentioned ``BranchPackage`` throws ~10 different ``ApiErrors``, some of them we might consider as bugs in this context, some of them not. To give some examples:

* ``DoubeBranchError``: Means the package already exists in the StagingProject. This can have  several reasons like somebody manually messed with the StagingProject, a different Request is already staged which updates the same package, the accept / cleanup did not finish properly. In this case we could handle this Error in a specific way like deleting the package and branching again or update the link or just ignoring this request. But this is not that simple and might need some more thoughts.
* ``InvalidProjectNameError`` this should for instance never happen and should be considered as a bug
* ``NotFoundError`` this might happen if somebody deletes the source package. What are we going to do in this case?

So I think this is a very simple but valid first implementation. Handling the different errors we could do in a follow up when we know what happens how often (errbit). But I agree that this might not the best user feedback / interface just removing the association again. We might need to think about how we can give the user a feedback of this background job like pasting a log somewhere in the project, but this just as an idea.

Reply:
Thanks for the explanation.

> DoubeBranchError: Means the package already exists in the StagingProject. [...] In this case we could handle this Error in a specific way like deleting the package and branching again or update the link or just ignoring this request. But this is not that simple and might need some more thoughts.

We could add a condition and check if the package already exists. :thinking:

> InvalidProjectNameError this should for instance never happen and should be considered as a bug

So it's nothing we would expect and thus don't have to cover it.

> NotFoundError this might happen if somebody deletes the source package. What are we going to do in this case?

Deleting the source package would result in revoking the request. Which means if we ever hit this exception here it would be a bug as well.

I would try to handle expected cases, like the package was already created, different from exceptions that we don't expect (which in most cases are bugs). That way it's much easier to spot the actual bug.

Reply:
The other issue I see is that it seems like we just remove staged requests, without telling users that this happened and why. So users might not notice at all that some request(s) were not processed or they realize that just by accident.

Reply:
> We could add a condition and check if the package already exists. thinking

And then? What are we going to do? Delete the package? Show an error message? This needs then happen everything in the controller ...

Btw these were just three examples I mentioned, there are many more. And we're still in the stage of implementing the skateboard :surfing_man:, not the Porsche :car: :wink: 


> The other issue I see is that it seems like we just remove staged requests, without telling users that this happened and why. So users might not notice at all that some request(s) were not processed or they realize that later on just by accident.

Yes, that's an issue and I already mentioned it in the limitations. We should have a way to give  the user feedback about the result of the staging job. Any good idea?

Reply:
> ChrisBr 6 hours ago  Member
>
>Btw these were just three examples I mentioned, there are many more. And we're still in the stage of implementing the skateboard 🏄, not the Porsche 🚗 😉

We are always implementing skateboards, aka [minimal viable product](https://blog.leanstack.com/minimum-viable-product-mvp-7e280b0b9418). But this isn't about the product anyway. It's about exception handling.
Right now we are mixing expected behaviour with bugs. Which will make it harder to spot actual bugs.
So if we can avoid or at least minimize this I would go for it. Let's ignore DoubeBranchError exceptions.

> Yes, that's an issue and I already mentioned it in the limitations. We should have a way to give the user feedback about the result of the staging job. Any good idea?

We would have to store the error messages within the delayed job. Either as part of the request or the staging project.

That's indeed something for another iteration.